### DISCUSSION TITLE
best practices for scripting

### DISCUSSION THREAD

**Discussion by laur89** [rockowitz/ddcutil#378] [Q&A] [answered]
Returning to ddcutil after few years and glad to see how far it's come! Detects both of the screens connected via docking station and marks phantom display appropriately. All out of the box (v2.1.2).

1. Want to write a simple script to modify brightness of laptop built-in and any number of connected screens. I understand built-in displays don't support ddc, so I suppose that is best delegated for other tools such as [brillo](https://gitlab.com/cameronnemo/brillo)? Or are there already standard patterns/methods that better address this requirement?

1. How about parsing `ddcutil` output? Unless I'm missing relevant part in manual, it doesn't support output format best suited for computer consumption? Guess I'm asking whether there's a better method for retrieving say brightness value than
    ```sh
    $ ddcutil getvcp --bus 8 10  | grep -Po 'current value\s*=\s*\K\d+'
    100
    ```

1. To keep script interaction time as fast as possible, it's best to address screens by the `--bus` identifier. I'm thinking caching `detect` output somewhere for quick retrieval of bus values, and update the cache on display connections/disconnections. Any issues with this approach?

1. And last, are you aware of scripts/programs that already do what I'm after to keep me from reinventing the wheel?
  **Edit:** wrote a simple python daemon to take care of brightness control that depends on ddcutil for external displays: https://github.com/laur89/bctl

**Comment by laur89** [comment]
@rockowitz sry for the ping, but if you have time I'd appreciate some input here.

**Comment by rockowitz** [comment, marked as answer]
I've twice written extended answers that seem to have gone lost.  So this one, I'm afraid, will be brief.

1) For a list of programs to control screen brightness, see this Arch Linux [page](https://wiki.archlinux.org/title/Backlight#External_monitors).  

2) See option ***--brief***

3) If you're writing a script for your own use, caching display information works.  As a released program, the devil is in the edge cases.  Unreleased option ***--enable-displays-cache*** was an attempt to do this, but reliably checking the edge cases negated the performance gain.  The shared library has part of an implementation for this, but the best that could be done in upcoming release 2.1.3 is to just tell the client that "something" has happened and leave it to the client to initiate complete display redetection.  See file src/ddc/ddcutil_watch_displays.c. 

4) See vdu_controls and this [gnome extension](https://extensions.gnome.org/extension/2645/brightness-control-using-ddcutil/).

**Comment by laur89** [comment]
1. already worked through arch wiki. All the alternatives - including brillo - depend on `ddcci-driver-linux-dkms` which does not appear to be working for me. No errors, but it just doesn't do the trick.
1. god damn, I swear I read the man page, but my comprehension must be lacking.
1. cheers, will read more about the option, but as the script would be only for my own machines I believe caching based on udev rules would work well enough.
1. huh, and written in js

Thanks as always!
